### PR TITLE
Add support for `#[default = "..."]` in methods as well

### LINKED ISSUE #435: #[default = "NULL"] doesn't work in `impl` block
The use of `#[default = "NULL"] ` in the following example, seems to be ignored. This works fine for individual functions, but not when the function is part of an `impl` block.

```rust
struct Person {
    pub name: String,
}

#[extendr]
impl Person {
    fn new(#[default = "NULL"] name: &str) -> Self {
        Self {
            name: name.to_string(),
        }
    }

    fn set_name(&mut self, name: &str) {
        self.name = name.to_string();
    }

    fn name(&self) -> &str {
        self.name.as_str()
    }
}
```

### PR DESCRIPTION
**Pull request by Ilia-Kosenkov**
Fixes #435.
Unify how functions' and methods' args are handled when wrappers are generated. Slightly rework how formal and actual arguments are generated from `Arg` metadata.

### CODE REVIEW

File: extendr-api/src/metadata.rs [hunk 1]
```diff
@@ -43,6 +43,37 @@ pub struct Metadata {
     pub impls: Vec<Impl>,
 }
 
+struct RArg {
+    name: String,
+    default: Option<&'static str>,
+}
+
+impl RArg {
```
Comment (multimeric) [inline extendr-api/src/metadata.rs:51]: I think it's a bit misleading to call this `RArg` when it's actually representing a *Rust* argument that will be converted into R.

File: extendr-api/src/metadata.rs [hunk 2]
```diff
@@ -215,25 +230,16 @@ fn write_method_wrapper(
...
+
+    // Skip a leading "self" argument.
+    // This is supplied by the environment.
+    let n_skip = r_args
+        .first()
+        .map_or(0, |a| if a.is_self() { 1 } else { 0 });
 
-    let formal_args = formal_args.join(", ");
-    let actual_args = actual_args.join(", ");
+    let formal_args = r_args.iter().skip(n_skip).map(|a| a.to_formal_arg());
```
Comment (multimeric) [inline extendr-api/src/metadata.rs:242]: Could we not express this more concisely by just `.filter()`ing this instead of using `.skip()`?

**Review by multimeric — APPROVED** [review]

File: extendr-api/src/metadata.rs (hunk 1 above)
Comment (Ilia-Kosenkov) [reply extendr-api/src/metadata.rs:51]: I originally tried to expand the functionality of `Arg`, but then realized it is used in many places. Here I want to introduce a struct to contain metadata of R arguments, mainly sanitized name and default value (if present). Hence the name. So it is not really *Rust* argument but *R* argument. 

Maybe we can improve on naming by utilizing the difference between *parameter* and *argument*. Here R parameters are names with optional default values that go into R function definition, and R arguments are actual arguments supplied to the `.Call()` instruction within that wrapper.

I am open to suggestions.

File: extendr-api/src/metadata.rs (hunk 2 above)
Comment (Ilia-Kosenkov) [reply extendr-api/src/metadata.rs:242]: I propose a middle ground in form of `.skip_while(|a| a.is_self())`. This communicates the idea that we want to drop the `self` argument found at the beginning of the argument list.

Comment (multimeric) [reply extendr-api/src/metadata.rs]: Sounds good to me.

File: extendr-api/src/metadata.rs (hunk 1 above)
Comment (multimeric) [reply extendr-api/src/metadata.rs:51]: Okay I take your point, I think this is a reasonable name.

**Pull request merged**